ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -877,6 +939,61 @@ MessageFormatFlags getOperationFlag() {
           : MessageFormatFlags.All;
     }
 
+    @Override
+    protected boolean maybeProcessDecryptCallbackAndCompleteOperation() {
+      if (decryptionRequired.get() && decryptCallbackResultInfo.resultAvailable) {
+        // in case of Metadata blob, only user-metadata needs decryption if the blob is encrypted
+        if (blobType == BlobType.MetadataBlob) {
+          if (decryptCallbackResultInfo.exception == null) {
+            logger.trace("Processing stored decryption callback result for Metadata blob {}", blobId);
+            chunkIdIterator = keys.listIterator();
```

Comment:
these are being done in `handleMetadataBlob()` - why does it have to be duplicated here? And why does this depend on whether `BlobInfo` is encrypted?

Reply:
yes. dataChunks are being used within poll() (check line 248). So, we shouldn't instantiate dataChunks unless metadata blob is completed. and incase of an encrypted metadata chunk, it is completed only when decryption result processing is complete which happens here. Will sync up with you offline to see if we can simplify this